source/Halibut.Tests/BadCertificatesTests.cs

```diff
@@ -0,0 +1,204 @@
+
+                countingService.GetCurrentValue().Should().Be(0, "With a bad certificate the request never should have been made");
+            }
+        }
+        
+        
+        [Test]
+        [LatestClientAndLatestServiceTestCases(testPolling: false, testWebSocket: false, testNetworkConditions: false)]
+        public async Task FailWhenListeningServicePresentsWrongCertificate(ClientAndServiceTestCase clientAndServiceTestCase)
+        {
+            var countingService = new CountingService();
+            using (var clientAndBuilder = await clientAndServiceTestCase.CreateTestCaseBuilder()
+                       .AsLatestClientAndLatestServiceBuilder()
+                       .WithClientTrustingTheWrongCertificate()
+                       .WithCountingService(countingService)
+                       .Build(CancellationToken))
+            {
+                var echo = clientAndBuilder.CreateClient<ICountingService>();
+                Assert.Throws<HalibutClientException>(() => echo.Increment())
+                    .Message.Should().Contain("" +
+                                              "We expected the server to present a certificate with the thumbprint 'EC32122053C6BFF582F8246F5697633D06F0F97F'. " +
+                                              "Instead, it presented a certificate with a thumbprint of '36F35047CE8B000CF4C671819A2DD1AFCDE3403D'");
+                countingService.GetCurrentValue().Should().Be(0, "With a bad certificate the request never should have been made");
+            }
+        }
+        
+        
+        
+        [Test]
+        [LatestClientAndLatestServiceTestCases(testListening: false, testNetworkConditions: false)]
+        public async Task FailWhenPollingServicePresentsWrongCertificate(ClientAndServiceTestCase clientAndServiceTestCase)
+        {
+            var countingService = new CountingService();
+            using (var clientAndBuilder = await clientAndServiceTestCase.CreateTestCaseBuilder()
+                       .AsLatestClientAndLatestServiceBuilder()
+                       .WithClientTrustingTheWrongCertificate()
+                       .WithCountingService(countingService)
+                       .RecordingClientLogs(out var serviceLoggers)
+                       .Build(CancellationToken))
+            {
+                var cts = new CancellationTokenSource();
+                var echo = clientAndBuilder.CreateClient<ICountingService, CancellationViaClientProxyFixture.IClientCountingService>(point =>
+                {
+                    point.PollingRequestQueueTimeout = TimeSpan.FromSeconds(10);
+                });
+                
+                var incrementCount = Task.Run(() => echo.Increment(new HalibutProxyRequestOptions(cts.Token)));
+
+                // Interestingly the message exchange error is logged to a non polling looking URL, perhaps because it has not been identified?
```

Comment:
Yes it hasn't `ReadRemoteIdentity` by this point so we don't know what subscription the polling tentacle is using!!